### PR TITLE
Verify the indices and provided values for index put meta

### PR DESCRIPTION
**Pull request by kundaMwiza**
- Verify index_put inputs when calling the meta function
- Add additional test inputs for index_put. Currently only indices provided for the first dimension are tested

Fixes #ISSUE_NUMBER

### CODE REVIEW

**Review by jansel — CHANGES_REQUESTED** [review]
Tests are failing

File: torch/_meta_registrations.py [hunk 1]
```diff
@@ -3643,8 +3645,52 @@ def meta_relu_(self):
     return self
 
 
+def check_index_put_inputs(self, indices, values, accumulate=False):
```
Comment (jansel) [inline torch/_meta_registrations.py:3648]: How does one verify this matches the non-meta behavior?  Is this based on some checking in the actual index_put kernel?

**Review by jansel — CHANGES_REQUESTED** [review]

File: test/functorch/test_ops.py [hunk 2]
```diff
@@ -939,6 +943,11 @@ def vjp_of_vjp(*args_and_cotangents):
         xfail('sparse.mm', 'reduce'),
         xfail('as_strided_scatter', ''),  # calls as_strided
         xfail('index_reduce', ''),  # .item() call
+        # RuntimeError: linearIndex.numel()*sliceSize*nElemBefore == expandedValue.numel()
+        # INTERNAL ASSERT FAILED at "/var/lib/jenkins/workspace/aten/src/ATen/native/cuda/Indexing.cu":496,
+        # please report a bug to PyTorch. number of flattened indices did not match number of elements
+        # in the value tensor: 20 vs 40
```
[the remark this answers is not in the document]
Comment (kundaMwiza) [reply test/functorch/test_ops.py:949]: These failures are due to the addition of new test cases for index_put. In particular with inputs that require broadcasting. Below all tests pass except for test_vmap_index_put_cuda. If the batching rules are independent of the device, the success of test_vmap_index_put_cpu should translate to a success for test_vmap_index_put_cuda. 

```python
import torch

dim_size = 5
batch_size = 2
accumulate=True

@pytest.fixture
def inputs():
    self = torch.randn((dim_size, dim_size))
    indices = torch.tensor([0, 1])
    values = torch.randn(1)
    return self, indices, values 
   
@pytest.mark.parametrize("device", ["cuda", "cpu"])
def test_index_put(device, inputs):
    return torch.index_put(inputs[0].to(device), [inputs[1].to(device)], inputs[2].to(device), accumulate=accumulate)

@pytest.mark.parametrize("device", ["cuda", "cpu"])
def test_vmap_index_put(device, inputs):

    def helper_fn(self, indices, values):
        return torch.index_put(self, indices, values, accumulate=accumulate) 

    self = inputs[0].expand((batch_size, dim_size, dim_size)).to(device)
    indices = [inputs[1].expand((batch_size, 2)).to(device)]
    values = inputs[2].expand((batch_size,)).to(device)
    torch.vmap(helper_fn)(self, indices, values)
```

File: torch/_meta_registrations.py (hunk 1 above)
Comment (kundaMwiza) [reply torch/_meta_registrations.py:3648]: Thats correct. The kernels for index.Tensor and index_put both use the same logic to verify the inputs and this logic has been replicated in Python for index.Tensor, see meta_index_Tensor in this file. check_index_put_inputs just adds some initial checking due to the difference between index_put and index.Tensor before calling meta_index_Tensor.

File: test/functorch/test_ops.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jansel) [reply test/functorch/test_ops.py:949]: Shouldn't invalid inputs not trigger internal asserts?

File: torch/_meta_registrations.py (hunk 1 above)
Comment (jansel) [reply torch/_meta_registrations.py:3648]: I am asking for some help here as a reviewer.  Reading this code I don't see how I'm supposed to verify this matches the C++ implementation.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_meta_registrations.py (hunk 1 above)
Comment (kundaMwiza) [reply torch/_meta_registrations.py:3648]: Both index.Tensor and index_put use the same logic (using the function make_info) when verifying the indicies and values.

The meta kernel for index.Tensor is [here](https://github.com/pytorch/pytorch/blob/93372455a73043332c16a71cb9dccdf3e0412a57/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L430) and the kernel calls the function [make_info](https://github.com/pytorch/pytorch/blob/93372455a73043332c16a71cb9dccdf3e0412a57/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L458C1-L459C1) to do most of the input verification.

Verification for the index_put inputs is also done by calling make_info. The index_put kernel can be found [here](https://github.com/pytorch/pytorch/blob/93372455a73043332c16a71cb9dccdf3e0412a57/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L708) and the call to make_info [further down](https://github.com/pytorch/pytorch/blob/93372455a73043332c16a71cb9dccdf3e0412a57/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L740). 

The make_info function is implemented [here](https://github.com/pytorch/pytorch/blob/93372455a73043332c16a71cb9dccdf3e0412a57/aten/src/ATen/native/TensorAdvancedIndexingUtils.h#L57)

The make_info function has been replicated in python: https://github.com/pytorch/pytorch/blob/93372455a73043332c16a71cb9dccdf3e0412a57/torch/_meta_registrations.py#L2786

Previously the python meta kernel for index_put did not verify that the indices and values are valid (e.g. running the checks in make_info). This PR enables that by calling into the replicated python logic for make_info

File: test/functorch/test_ops.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kundaMwiza) [reply test/functorch/test_ops.py:949]: I'm not sure of the restrictions of torch.vmap and CUDA. The above inputs work on the CPU so I assume the inputs are also valid for CUDA.

**Review by jansel — APPROVED** [review]

File: test/onnx/test_fx_op_consistency.py [hunk 3]
```diff
@@ -422,6 +422,28 @@ def skip_torchlib_forward_compatibility(
         dtypes=(torch.uint8, torch.int8, torch.int16,),
         reason=onnx_test_common.reason_onnx_script_does_not_support("Add", "int8, int16"),
     ),
+    xfail(
+        "index_put",
+        matcher=lambda sample: len(sample.args[0]) == 1 and sample.args[0][0].dim() == 0,
+        reason=(
+            "fixme: "
+            "[ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Non-zero status code "
```
[the remark this answers is not in the document]
Comment (thiagocrepaldi) [reply test/onnx/test_fx_op_consistency.py:430]: @justinchuby @titaiwangms FYI

**Pull request closed without merging**